common/lib/capa/capa/responsetypes.py

```diff
@@ -831,16 +841,110 @@ def assign_choice_names(self):
+        elif halves_error_count == 1 and len(all_choices) > 2:
+            return_grade = round(self.get_max_score() / 2.0, 2)
+            return CorrectMap(self.answer_id, correctness='partially-correct', npoints=return_grade)
+        elif halves_error_count == 2 and len(all_choices) > 4:
+            return_grade = round(self.get_max_score() / 4.0, 2)
+            return CorrectMap(self.answer_id, correctness='partially-correct', npoints=return_grade)
+        else:
+            return CorrectMap(self.answer_id, 'incorrect')
+
     def get_score(self, student_answers):
 
+        # Setting up answer lists:
+        #  all_choices: the full list of checkboxes
+        #  student_answer: what the student actually chose (note no "s")
+        #  student_non_answer: what they didn't choose
+        #  self.correct_choices: boxes that should be checked
+        #  self.incorrect_choices: boxes that should NOT be checked
+
+        all_choices = self.correct_choices.union(self.incorrect_choices)
+
         student_answer = student_answers.get(self.answer_id, [])
 
         if not isinstance(student_answer, list):
             student_answer = [student_answer]
 
+        # "None apply" should really be a valid choice for "check all that apply",
+        # but it throws an error if all the checks are blank.
         no_empty_answer = student_answer != []
+
         student_answer = set(student_answer)
 
+        student_non_answers = all_choices - student_answer
+
+        if not no_empty_answer:
+            return CorrectMap(self.answer_id, 'incorrect')
+
+        # This below checks to see whether we're using an alternate grading scheme.
+        #  Set partial_credit="false" (or remove it) to require an exact answer for any credit.
+        #  Set partial_credit="EDC" to count each choice for equal points (Every Decision Counts).
+        #  Set partial_credit="halves" to take half credit off for each error.
+
+        tree = self.xml
+        problem_xml = tree.xpath('.')
+
+        # Partial credit type - only one type at a time right now.
+        credit_type = problem_xml[0].get('partial_credit', default=False)
+
+        try:
+            credit_type = str(credit_type).lower().strip()
+        except ValueError:
+            _ = self.capa_system.i18n.ugettext
+            # Translators: 'partial_credit' is an attribute name and should not be translated.
+            # 'EDC' and 'halves' and 'false' should also not be translated.
+            msg = _("partial_credit value should be one of 'EDC', 'halves', or 'false'.")
+            raise LoncapaProblemError(msg)
+
+        if credit_type == 'false':
+            pass
+        elif credit_type == 'halves':
+            return self.grade_via_halves(all_choices, student_answer, student_non_answers)
+        elif credit_type == 'edc':
+            return self.grade_via_edc(all_choices, student_answer, student_non_answers)
+
```

Comment:
There are two issues here: 

1. This chunk of code should be generic. The schemes are relisted many times. DRY.
2. It fails to handle invalid strings well. The only time it fails is for non-strings. 

To put it in pseudocode, I'd rather have this defined once in a map: 

     graders = {'edc':self.grade_via_edc, 'halves': grade_via_halves, 'false': grade_via_[some good name]'}
     if credit_type not in graders:
        raise "partial credit should be one of: ", ",".join(graders)
     graders[credit_type](all_choices, student_answer, student_non_answers)

(the rest of the function -- which defines one particular grading scheme -- should then be broken out into its own grader function). 

Ideally, this code would be globally generic in ResponseType. Each individual child (ChoiceResponse and similar) could just define a set of grader functions, and either the map of graders, or even just have those graders take a decorator: 

    \@grading_scheme(['edc', 'every-decision-counts'])
   def grade_via_edc(...):

I could also then define graders outside of the particular problem (or in CapaResponse) and include them either in the base class (if generic to all) or in the child class. E.g.: 

    \@grading_scheme(['always-correct'])
    def grade_always_correct(...):
        return CorrectMap(self.answer_id, ''correct")

This would be much less code overall, contain less repetition, more readable, more extensible, .